File: clients/apps/nodejs/test/ta02-duplicate-deposits.mjs
import { expect } from 'chai'; 
import client_config from '../client_config.js';
import mercurynodejslib from 'mercurynodejslib';
import { CoinStatus } from 'mercurynodejslib/coin_enum.js';
import { createWallet, removeDatabase, getnewaddress, generateBlock, depositCoin } from './test-utils.mjs';

describe('TA02 - Duplicated Deposits', function() {

    context('Withdraw Flow', () => {
        it('should complete successfully', async () => {

            const clientConfig = client_config.load();
            await removeDatabase(clientConfig);

            let wallet_1_name = "w1";
            let wallet_2_name = "w2";
            let wallet1 = await createWallet(clientConfig, wallet_1_name);
            let wallet2 = await createWallet(clientConfig, wallet_2_name);

            const token = await mercurynodejslib.newToken(clientConfig, wallet1.name);
            const tokenId = token.token_id;

            const amount1 = 10000;
            const depositInfo = await mercurynodejslib.getDepositBitcoinAddress(clientConfig, wallet1.name, amount1);

            await depositCoin(depositInfo.deposit_address, amount1);

            let amount2 = 20000;
            await depositCoin(depositInfo.deposit_address, amount2);

            const coreWalletAddress = await getnewaddress();

            await generateBlock(clientConfig.confirmationTarget, coreWalletAddress);

            const listCoins = await mercurynodejslib.listStatecoins(clientConfig, wallet1.name);

            expect(listCoins.length).to.equal(2);

            const newCoin = listCoins.find(coin => 
                coin.aggregated_address == depositInfo.deposit_address && 
                coin.status == CoinStatus.CONFIRMED
            );
              
            const duplicatedCoin = listCoins.find(coin => 
                coin.aggregated_address == depositInfo.deposit_address && 
                coin.status == CoinStatus.DUPLICATED
            );

            expect(newCoin).to.not.be.null;
            expect(duplicatedCoin).to.not.be.null;

            expect(newCoin.duplicate_index).to.equal(0);
            expect(duplicatedCoin.duplicate_index).to.equal(1);

            const transferAddress = await mercurynodejslib.newTransferAddress(clientConfig, wallet2.name, null);

            try {
                await mercurynodejslib.transferSend(clientConfig, wallet1.name, newCoin.statechain_id, transferAddress.transfer_receive, false, null);
            } catch (error) {
                expect(error.message).to.include("Coin is duplicated. If you want to proceed, use the command '--force, -f' option. " + 
                    "You will no longer be able to move other duplicate coins with the same statechain_id and this will cause PERMANENT LOSS of these duplicate coin funds.");
            }

            const withdrawAddress = "bcrt1qn5948np2j8t68xgpceneg3ua4wcwhwrsqj8scv";

            let txid = await mercurynodejslib.withdrawCoin(clientConfig, wallet1.name, duplicatedCoin.statechain_id, withdrawAddress, null, 1);

            expect(txid).to.be.string;

            try {
                await mercurynodejslib.transferSend(clientConfig, wallet1.name, newCoin.statechain_id, transferAddress.transfer_receive, false, null);
            } catch (error) {
                expect(error.message).to.include("There have been withdrawals of other coins with this same statechain_id (possibly duplicates). " +
                    "This transfer cannot be performed because the recipient would reject it due to the difference in signature count. This coin can be withdrawn, however.");
            }
            
            txid = await mercurynodejslib.withdrawCoin(clientConfig, wallet1.name, duplicatedCoin.statechain_id, withdrawAddress, null, 0);

            expect(txid).to.be.string;
        })
    }),

    context('Transfer Flow', () => {
        it('should complete successfully', async () => {

            const clientConfig = client_config.load();
            await removeDatabase(clientConfig);

            let wallet_1_name = "w1";
            let wallet_2_name = "w2";
            let wallet1 = await createWallet(clientConfig, wallet_1_name);
            let wallet2 = await createWallet(clientConfig, wallet_2_name);

            const token = await mercurynodejslib.newToken(clientConfig, wallet1.name);
            const tokenId = token.token_id;

            const amount1 = 10000;
            const depositInfo = await mercurynodejslib.getDepositBitcoinAddress(clientConfig, wallet1.name, amount1);

            await depositCoin(depositInfo.deposit_address, amount1);

            let amount2 = 20000;
            await depositCoin(depositInfo.deposit_address, amount2);

            const coreWalletAddress = await getnewaddress();

            await generateBlock(clientConfig.confirmationTarget, coreWalletAddress);

            let listCoins = await mercurynodejslib.listStatecoins(clientConfig, wallet1.name);

            expect(listCoins.length).to.equal(2);

            const newCoin = listCoins.find(coin => 
                coin.aggregated_address == depositInfo.deposit_address && 
                coin.status == CoinStatus.CONFIRMED
            );
              
            let duplicatedCoin = listCoins.find(coin => 
                coin.aggregated_address == depositInfo.deposit_address && 
                coin.status == CoinStatus.DUPLICATED
            );

            expect(newCoin).to.not.be.null;
            expect(duplicatedCoin).to.not.be.null;

            expect(newCoin.duplicate_index).to.equal(0);
            expect(duplicatedCoin.duplicate_index).to.equal(1);

            const transferAddress = await mercurynodejslib.newTransferAddress(clientConfig, wallet2.name, null);

            let result = await mercurynodejslib.transferSend(clientConfig, wallet1.name, newCoin.statechain_id, transferAddress.transfer_receive, true, null);
            expect(result).to.have.property('statechain_id');

            let transferReceiveResult = await mercurynodejslib.transferReceive(clientConfig, wallet2.name);

            expect(transferReceiveResult.receivedStatechainIds).contains(newCoin.statechain_id);
            expect(transferReceiveResult.receivedStatechainIds.length).to.equal(1);

            listCoins = await mercurynodejslib.listStatecoins(clientConfig, wallet1.name);

            const transferredCoin = listCoins.find(coin => 
                coin.aggregated_address === depositInfo.deposit_address && 
                coin.status === CoinStatus.TRANSFERRED
            );
              
            duplicatedCoin = listCoins.find(coin => 
                coin.aggregated_address === depositInfo.deposit_address && 
                coin.status === CoinStatus.DUPLICATED
            );

            expect(transferredCoin).to.not.be.null;
            expect(transferredCoin.duplicate_index).to.equal(0);

            expect(duplicatedCoin).to.not.be.null;
            expect(duplicatedCoin.duplicate_index).to.equal(1);

            try {
                const withdrawAddress = "bcrt1qn5948np2j8t68xgpceneg3ua4wcwhwrsqj8scv";
                await mercurynodejslib.withdrawCoin(clientConfig, wallet1.name, duplicatedCoin.statechain_id, withdrawAddress, null, 1);
            } catch (error) {
                expect(error.response.status).to.equal(401);
                expect(error.response.data.message).to.equal("Signature does not match authentication key.");
            }
        })
    })
})

File: clients/tests/rust/src/ta02_duplicate_deposits.rs
use std::{env, process::Command, thread, time::Duration};

use anyhow::{Result, Ok};
use mercuryrustlib::{client_config::ClientConfig, CoinStatus, Wallet};

use crate::{bitcoin_core, electrs};

async fn withdraw_flow(client_config: &ClientConfig, wallet1: &Wallet, wallet2: &Wallet)  -> Result<()> {

    let amount = 1000;

    let token_id = mercuryrustlib::deposit::get_token(client_config).await?;

    let deposit_address = mercuryrustlib::deposit::get_deposit_bitcoin_address(&client_config, &wallet1.name, &token_id, amount).await?;

    let _ = bitcoin_core::sendtoaddress(amount, &deposit_address)?;

    let core_wallet_address = bitcoin_core::getnewaddress()?;
    let remaining_blocks = client_config.confirmation_target;
    let _ = bitcoin_core::generatetoaddress(remaining_blocks, &core_wallet_address)?;

    // It appears that Electrs takes a few seconds to index the transaction
    let mut is_tx_indexed = false;

    while !is_tx_indexed {
        is_tx_indexed = electrs::check_address(client_config, &deposit_address, amount).await?;
        thread::sleep(Duration::from_secs(1));
    }

    mercuryrustlib::coin_status::update_coins(&client_config, &wallet1.name).await?;
    let wallet: mercuryrustlib::Wallet = mercuryrustlib::sqlite_manager::get_wallet(&client_config.pool, &wallet1.name).await?;
    let new_coin = wallet.coins.iter().find(|&coin| coin.aggregated_address == Some(deposit_address.clone())).unwrap();

    assert!(new_coin.status == CoinStatus::CONFIRMED);

    let amount = 2000;

    let _ = bitcoin_core::sendtoaddress(amount, &deposit_address)?;

    let _ = bitcoin_core::generatetoaddress(remaining_blocks, &core_wallet_address)?;

    let mut is_tx_indexed = false;

    while !is_tx_indexed {
        is_tx_indexed = electrs::check_address(client_config, &deposit_address, amount).await?;
        thread::sleep(Duration::from_secs(1));
    }

    mercuryrustlib::coin_status::update_coins(&client_config, &wallet1.name).await?;
    let wallet1: mercuryrustlib::Wallet = mercuryrustlib::sqlite_manager::get_wallet(&client_config.pool, &wallet1.name).await?;

    let new_coin = wallet1.coins.iter().find(|&coin| coin.aggregated_address == Some(deposit_address.clone()) && coin.status == CoinStatus::CONFIRMED);
    let duplicated_coin = wallet1.coins.iter().find(|&coin| coin.aggregated_address == Some(deposit_address.clone()) && coin.status == CoinStatus::DUPLICATED);

    assert!(new_coin.is_some());
    assert!(duplicated_coin.is_some());

    let new_coin = new_coin.unwrap();
    let duplicated_coin = duplicated_coin.unwrap();

    assert!(new_coin.duplicate_index == 0);
    assert!(duplicated_coin.duplicate_index == 1);

    let statechain_id = new_coin.statechain_id.as_ref().unwrap();

    let wallet2_transfer_adress = mercuryrustlib::transfer_receiver::new_transfer_address(&client_config, &wallet2.name).await?;

    let batch_id = None;

    let force_send = false;

    let result = mercuryrustlib::transfer_sender::execute(&client_config, &wallet2_transfer_adress, &wallet1.name, statechain_id, force_send, None, batch_id.clone()).await;

    assert!(result.is_err());

    let error_msg = result.err().unwrap().to_string();

    assert!(error_msg == "Coin is duplicated. If you want to proceed, use the command '--force, -f' option. \
        You will no longer be able to move other duplicate coins with the same statechain_id and this will cause PERMANENT LOSS of these duplicate coin funds.");

    let fee_rate = None;

    let result = mercuryrustlib::withdraw::execute(&client_config, &wallet1.name, statechain_id, &core_wallet_address, fee_rate, Some(1)).await;

    assert!(result.is_ok());

    mercuryrustlib::coin_status::update_coins(&client_config, &wallet1.name).await?;

    let result = mercuryrustlib::transfer_sender::execute(&client_config, &wallet2_transfer_adress, &wallet1.name, statechain_id, force_send, None, batch_id).await;

    assert!(result.is_err());

    let error_msg = result.err().unwrap().to_string();

    assert!(error_msg == "There have been withdrawals of other coins with this same statechain_id (possibly duplicates).\
        This transfer cannot be performed because the recipient would reject it due to the difference in signature count.\
        This coin can be withdrawn, however.");

    let result = mercuryrustlib::withdraw::execute(&client_config, &wallet1.name, statechain_id, &core_wallet_address, fee_rate, None).await;

    assert!(result.is_ok());
    
    Ok(())
}

async fn transfer_flow(client_config: &ClientConfig, wallet1: &Wallet, wallet2: &Wallet)  -> Result<()> {

    let amount = 1000;

    let token_id = mercuryrustlib::deposit::get_token(client_config).await?;

    let deposit_address = mercuryrustlib::deposit::get_deposit_bitcoin_address(&client_config, &wallet1.name, &token_id, amount).await?;

    let _ = bitcoin_core::sendtoaddress(amount, &deposit_address)?;

    let core_wallet_address = bitcoin_core::getnewaddress()?;
    let remaining_blocks = client_config.confirmation_target;
    let _ = bitcoin_core::generatetoaddress(remaining_blocks, &core_wallet_address)?;

    // It appears that Electrs takes a few seconds to index the transaction
    let mut is_tx_indexed = false;

    while !is_tx_indexed {
        is_tx_indexed = electrs::check_address(client_config, &deposit_address, amount).await?;
        thread::sleep(Duration::from_secs(1));
    }

    mercuryrustlib::coin_status::update_coins(&client_config, &wallet1.name).await?;
    let wallet: mercuryrustlib::Wallet = mercuryrustlib::sqlite_manager::get_wallet(&client_config.pool, &wallet1.name).await?;
    let new_coin = wallet.coins.iter().find(|&coin| coin.aggregated_address == Some(deposit_address.clone())).unwrap();

    assert!(new_coin.status == CoinStatus::CONFIRMED);

    let amount = 2000;

    let _ = bitcoin_core::sendtoaddress(amount, &deposit_address)?;

    let _ = bitcoin_core::generatetoaddress(remaining_blocks, &core_wallet_address)?;

    let mut is_tx_indexed = false;

    while !is_tx_indexed {
        is_tx_indexed = electrs::check_address(client_config, &deposit_address, amount).await?;
        thread::sleep(Duration::from_secs(1));
    }

    mercuryrustlib::coin_status::update_coins(&client_config, &wallet1.name).await?;
    let wallet1: mercuryrustlib::Wallet = mercuryrustlib::sqlite_manager::get_wallet(&client_config.pool, &wallet1.name).await?;

    let new_coin = wallet1.coins.iter().find(|&coin| coin.aggregated_address == Some(deposit_address.clone()) && coin.status == CoinStatus::CONFIRMED);
    let duplicated_coin = wallet1.coins.iter().find(|&coin| coin.aggregated_address == Some(deposit_address.clone()) && coin.status == CoinStatus::DUPLICATED);

    assert!(new_coin.is_some());
    assert!(duplicated_coin.is_some());

    let new_coin = new_coin.unwrap();
    let duplicated_coin = duplicated_coin.unwrap();

    assert!(new_coin.duplicate_index == 0);
    assert!(duplicated_coin.duplicate_index == 1);

    let statechain_id = new_coin.statechain_id.as_ref().unwrap();

    let wallet2_transfer_adress = mercuryrustlib::transfer_receiver::new_transfer_address(&client_config, &wallet2.name).await?;

    let batch_id = None;

    let force_send = true;

    let result = mercuryrustlib::transfer_sender::execute(&client_config, &wallet2_transfer_adress, &wallet1.name, statechain_id, force_send, None, batch_id.clone()).await;

    assert!(result.is_ok());
    
    let transfer_receive_result = mercuryrustlib::transfer_receiver::execute(&client_config, &wallet2.name).await?;
    let received_statechain_ids = transfer_receive_result.received_statechain_ids;

    assert!(received_statechain_ids.contains(&statechain_id.to_string()));
    assert!(received_statechain_ids.len() == 1);

    mercuryrustlib::coin_status::update_coins(&client_config, &wallet1.name).await?;
    let wallet1: mercuryrustlib::Wallet = mercuryrustlib::sqlite_manager::get_wallet(&client_config.pool, &wallet1.name).await?;

    let transferred_coin = wallet1.coins.iter().find(|&coin| coin.aggregated_address == Some(deposit_address.clone()) && coin.status == CoinStatus::TRANSFERRED);
    let duplicated_coin = wallet1.coins.iter().find(|&coin| coin.aggregated_address == Some(deposit_address.clone()) && coin.status == CoinStatus::DUPLICATED);

    assert!(transferred_coin.is_some());
    assert!(duplicated_coin.is_some());

    let transferred_coin = transferred_coin.unwrap();
    let duplicated_coin = duplicated_coin.unwrap();

    assert!(transferred_coin.duplicate_index == 0);
    assert!(duplicated_coin.duplicate_index == 1);

    let fee_rate = None;

    let result = mercuryrustlib::withdraw::execute(&client_config, &wallet1.name, statechain_id, &core_wallet_address, fee_rate, Some(1)).await;

    assert!(result.is_err());

    let error_msg = result.err().unwrap().to_string();

    assert!(error_msg == "Signature does not match authentication key.");



    Ok(())
}

pub async fn execute() -> Result<()> {

    let _ = Command::new("rm").arg("wallet.db").arg("wallet.db-shm").arg("wallet.db-wal").output().expect("failed to execute process");

    env::set_var("ML_NETWORK", "regtest");

    let client_config = mercuryrustlib::client_config::load().await;

    let wallet1 = mercuryrustlib::wallet::create_wallet(
        "wallet1", 
        &client_config).await?;

    mercuryrustlib::sqlite_manager::insert_wallet(&client_config.pool, &wallet1).await?;

    let wallet2 = mercuryrustlib::wallet::create_wallet(
        "wallet2", 
        &client_config).await?;

    mercuryrustlib::sqlite_manager::insert_wallet(&client_config.pool, &wallet2).await?;

    withdraw_flow(&client_config, &wallet1, &wallet2).await?;
    transfer_flow(&client_config, &wallet1, &wallet2).await?;

    println!("TA02 - Test \"Multiple Deposits in the Same Adress\" completed successfully");

    Ok(())
}